src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingExtensionMethods.cs

```diff
@@ -0,0 +1,132 @@
+                    string.Format(
+                        "Loaded SMO urn object count {0} for type {1}, urns: {2}",
+                        objectType,
+                        databaseObjectsOfType.Count(),
+                        string.Join(", ", databaseObjectsOfType.Select(p => p.Value))));
+
+                foreach (KeyValuePair<string, string> databaseObjectOfType in databaseObjectsOfType)
+                {
+                    if (!serverAndDatabaseInitialized)
+                    {
+                        Urn urn = new Urn(databaseObjectOfType.Value);
+                        serverName = urn.GetNameForType("Server");
+                        databaseName = urn.GetNameForType("Database");
+                        serverAndDatabaseInitialized = true;
+                    }
+
+                    databaseObjects.Add(new Urn(databaseObjectOfType.Value).ToScriptingObject());
+                }
+            }
+
+            return databaseObjects;
+        }
+
+        public static bool IsOperationCanceledException(this Exception e)
+        {
+            Exception current = e;
+            while (current != null)
+            {
+                if (current is OperationCanceledException)
+                {
+                    return true;
+                }
+
+                current = current.InnerException;
+            }
+
+            return false;
+        }
+
+        public static Urn ToUrn(this ScriptingObject scriptingObject, string server, string database)
+        {
+            if (scriptingObject == null)
+            {
+                throw new ArgumentNullException("scriptingObjectc");
+            }
+
+            if (string.IsNullOrWhiteSpace(server))
+            {
+                throw new ArgumentException("Parameter must have a value", "server");
+            }
+
+            if (string.IsNullOrWhiteSpace(database))
+            {
+                throw new ArgumentException("Parameter must have a value", "database");
+            }
+
+            if (string.IsNullOrWhiteSpace(scriptingObject.Name))
+            {
+                throw new ArgumentException("Property scriptingObject.Name must have a value", "scriptingObject");
+            }
+
+            if (string.IsNullOrWhiteSpace(scriptingObject.Type))
+            {
+                throw new ArgumentException("Property scriptingObject.Type must have a value", "scriptingObject");
+            }
+
+            string urn = string.Format(
+                "Server[@Name='{0}']/Database[@Name='{1}']/{2}[@Name='{3}' {4}]",
+                server,
+                database,
+                scriptingObject.Type,
+                scriptingObject.Name,
+                scriptingObject.Schema != null ? string.Format("and @Schema = '{0}'", scriptingObject.Schema) : string.Empty);
+
+            return new Urn(urn);
+        }
+
+        public static ScriptingObject ToScriptingObject(this Urn urn)
+        {
+            if (urn == null)
```

Comment:
>urn [](start = 16, length = 3)

Please use Validate.IsNotNull(nameof(urn), urn) as this is our standard pattern elsewhere